tools/crm_mon.c

```diff
@@ -3652,47 +4156,111 @@ mon_refresh_display(gpointer user_data)
     switch (output_format) {
         case mon_output_html:
         case mon_output_cgi:
-            if (print_html_status(&data_set, output_filename) != 0) {
+            if (print_html_status(&data_set, output_filename, stonith_history) != 0) {
                 fprintf(stderr, "Critical: Unable to output html file\n");
                 clean_up(CRM_EX_CANTCREAT);
             }
             break;
 
         case mon_output_xml:
-            print_xml_status(&data_set);
+            print_xml_status(&data_set, stonith_history);
             break;
 
         case mon_output_monitor:
-            print_simple_status(&data_set);
+            print_simple_status(&data_set, stonith_history);
             if (has_warnings) {
                 clean_up(MON_STATUS_WARN);
             }
             break;
 
         case mon_output_plain:
         case mon_output_console:
-            print_status(&data_set);
+            print_status(&data_set, stonith_history);
             break;
 
         case mon_output_none:
             break;
     }
 
+    stonith_history_free(stonith_history);
+    stonith_history = NULL;
     cleanup_alloc_calculations(&data_set);
+
+    /* poll for pending fence-actions as we don't get notifications so far */
+    if ((!one_shot) && (show & mon_show_fence_history)) {
+        kick_refresh(FALSE);
+    }
     return TRUE;
 }
 
 void
-mon_st_callback(stonith_t * st, stonith_event_t * e)
+mon_st_callback_event(stonith_t * st, stonith_event_t * e)
 {
-    char *desc = crm_strdup_printf("Operation %s requested by %s for peer %s: %s (ref=%s)",
-                                 e->operation, e->origin, e->target, pcmk_strerror(e->result),
-                                 e->id);
+    if (st->state == stonith_disconnected) {
+        st->state = stonith_disconnected;
+        /* disconnect cib as well and have everything reconnect */
+        mon_cib_connection_destroy(NULL);
+    } else {
```

Comment:
looks like this could be else if (external_agent)

Alternatively, maybe we don't even need to register this callback unless external_agent is used?

Reply:
yep looks as if. I guess that is a leftover from you removing smtp.